### PR TITLE
Cleanup setuppy

### PR DESCRIPTION
**Pull request by himkt**
To cleanup `setup.py`, I propose the following refactoring:

- remove `experimental`. It's not used
- unify `document` and `doctest`. They always co-occur
- unify `tests` and `optional` and create `develop`
- unify `testing` and `integration`. A lot of overlap

## Motivation
🔗 https://github.com/optuna/optuna/issues/1576

### CODE REVIEW

File: setup.py
Comment (himkt) [inline setup.py:56]: `virtualenv` is removed since it is builtin.

https://github.com/optuna/optuna/pull/3517/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L174-L177

File: setup.py
Comment (himkt) [inline setup.py:219]: `tests_require` is deprecated.

https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies
https://github.com/pypa/setuptools/issues/1684

File: setup.py
Comment (himkt) [inline setup.py:141]: `tests` and `optional` are always used simultaneously to make me think of simply unifying. @not522 

https://cs.github.com/optuna/optuna?q=tests%20path%3A%2F%5E.github%5C%2Fworkflows%5C%2F%2F

File: setup.py
Comment (himkt) [inline setup.py:140]: `tests` and `optional` are always used simultaneously to make me think of simply unifying. @not522 

https://cs.github.com/optuna/optuna?q=tests%20path%3A%2F%5E.github%5C%2Fworkflows%5C%2F%2F

File: setup.py
Comment (nzw0301) [inline setup.py:86]: duplicated
```suggestion
```

File: setup.py
Comment (not522) [reply setup.py:140]: I prefer to separate tests depending on whether they require additional libraries. E.g.,

```
pip install .[tests]
pytest tests -m "not optional"
pip install .[optional]
pytest tests -m "optional"
```

It will prevent accidentally importing optional libraries as default.

Comment (himkt) [reply setup.py]: [5ad5e76](https://github.com/optuna/optuna/pull/3517/commits/5ad5e7634813c0e1d37f0045c93263bc6c8b2581)
Restored `test` (plural -> singular for the consistency with other extras) and `optional`. 🙌

**Comment by himkt** [comment]
Sorry for the little bit late response. I revised the PR, PTAL @not522 @keisuke-umezawa (cc. @nzw0301).

File: setup.py
Comment (not522) [inline setup.py:69]: duplicated

Comment (himkt) [reply setup.py]: Sorry :bow: I removed the duplication in 8251ff1.

BTW, matplotlib v3.0.0 was released in Sep 18, 2018 (3+ years ago). Can we remove the version constraint now?

Comment (not522) [reply setup.py]: I suppose these install options are only used by Optuna developers or CI, and we can remove the lower constraint.

Comment (himkt) [reply setup.py]: 👍 , removed in [461afdd](https://github.com/optuna/optuna/pull/3517/commits/461afdd4df45e3f70c1fb9de7ba0e2e6d8bb03fd).

**Review by not522 — APPROVED** [review]
LGTM!

**Comment by keisuke-umezawa** [comment]
Thank you for cleaning up it! We also need to update the following files.

* https://github.com/optuna/optuna/blob/master/tox.ini#L14

**Comment by himkt** [comment]
You got the point, thank you so much for the careful review! ❤️
I updated `tox.ini` in 7685b30.

**Comment by himkt** [comment]
I'll revise the PR after https://github.com/optuna/optuna/pull/3510 being merged.
Let me make this PR draft for now.

**Comment by nzw0301** [comment]
Since this PR has almost two approvals, so I'm happy to resolve the forthcoming conflict in #3510 after merging this PR onto the master branch.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! I also merged #3510.

**Comment by himkt** [comment]
@not522 @keisuke-umezawa (cc. @nzw0301)

I fixed the conflict with the master branch.
I'm so sorry to bother you, but please review the PR again.

**Review by not522 — APPROVED** [review]
LGTM, thank you!

**Review by keisuke-umezawa — APPROVED** [review]
LGTM!

File: setup.py
Comment (nzw0301) [inline setup.py:128]: This reverted the change to relax the version constraint of `fakeredis`.

```
            "fakeredis<=1.7.1; python_version<'3.8'",
            "fakeredis ; python_version>='3.8'",
```

**Pull request merged**
